types/wlr_surface.c

```diff
@@ -149,15 +149,20 @@ static void surface_state_finalize(struct wlr_surface *surface,
 		}
 	}
 
-	int width = state->buffer_width / state->scale;
-	int height = state->buffer_height / state->scale;
-	if ((state->transform & WL_OUTPUT_TRANSFORM_90) != 0) {
-		int tmp = width;
-		width = height;
-		height = tmp;
+	if (state->viewport.has_dst && state->buffer_resource != NULL) {
+		state->width = state->viewport.dst_width;
+		state->height = state->viewport.dst_height;
+	} else {
```

Comment:
If the surface is only cropped, then the surface width and the surface height should be `state->viewport.src_width` and `state->viewport.src_height`, respectively.